libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/coordinator/NavigationStateManager.kt

```diff
@@ -0,0 +1,88 @@
+package com.mapbox.navigation.dropin.coordinator
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.navigation.base.trip.model.RouteProgressState
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.dropin.DropInNavigationViewContext
+import com.mapbox.navigation.dropin.component.navigationstate.NavigationState
+import com.mapbox.navigation.dropin.extensions.flowOnFinalDestinationArrival
+import com.mapbox.navigation.dropin.extensions.flowRoutesUpdated
+import com.mapbox.navigation.dropin.lifecycle.UIComponent
+import com.mapbox.navigation.dropin.model.Destination
+import kotlinx.coroutines.flow.combine
+
+/**
+ * Class that manages NavigationState as following:
+ *
+ * ```
+ *  [ FreeDrive ]━━( destination != null )━━━━━━>>[ RoutePreview ]<<━━━━━━━━━━━━┓
+ *       ^         ( & routes not empty  )               ┃                      ┃
+ *       ┃                                  (activeNavigationStarted=true)      ┃
+ *       ┃                                               ┃                      ┃
+ *       ┃                                               ┃         (activeNavigationStarted=false)
+ *       ┣━━━━( destination == null )━━━━┓               v                      ┃
+ *       ┗━━━━( routes empty        )━━━━╋━━━━━[ ActiveNavigation ]━━━━━━━━━━━━━┫
+ *                                       ┃               ┃                      ┃
+ *                                       ┃  (onFinalDestinationArrival())       ┃
+ *                                       ┃               ┃                      ┃
+ *                                       ┃               v                      ┃
+ *                                       ┗━━━━━━━━━━[ Arrival ]━━━━━━━━━━━━━━━━━┛
+ * ```
+ */
+internal class NavigationStateManager(
```

Comment:
How do you envision developer being able to initialize the drop-in UI in any of the predefined states? It's not something that has to be implemented in this PR but we should think how to leave that door open.

Reply:
We didn't talk about how will host application interact with the Drop In UI, yet. 
I am still waiting for the "Technical document" (see my questions at the end of the "FAQ: Drop-In UI for Navigation SDK" doc).

Here are a couple of ways we can achieve this:
1. Expose single public method via DropInNavigationView
```kotlin
// free drive
navVIew.setInitialStateFreeDrive()
// free drive with POI
navVIew.setInitialStateFreeDrive(destination: Destination)

// route preview (we will fetch routes)
navView.setInitialStatePreview(destination: Destination)

// active guidance (we will fetch routes and start navigation session) 
navVIew.setInitialStateActiveGuidance(destination: Destination)
```

2.  Expose a mediator/controller/manager that can be used to retrieve current state and control nav UI.
```kotlin
navView.controller.setDestination(destination)
navView.controller.getNavigationState()
navView.controller.showRoutePreview(destination)
navView.controller.setVoiceInstructionsEnabled(enabled)
```